Makefile

```diff
@@ -274,71 +58,181 @@ OCTP_DEPS = \
 	$(DIST_DIR)/lib/libfontconfig.a \
 	$(DIST_DIR)/lib/libass.a
 
-# Require a patch to fix some errors
+WORKER_DEPS = \
+	$(DIST_DIR)/lib/libsubtitles-octopus.a \
+	$(OCTP_DEPS)
+
+all: subtitleoctopus
+
+subtitleoctopus: dist
+
+dist_zip: subtitleoctopus zip
+
+$(BASE_DIR)/build/lib/created:
+	mkdir -p build/lib
+	touch $(BASE_DIR)/build/lib/created
+
+include $(BASE_DIR)/build/m4/fribidi.mk
+include $(BASE_DIR)/build/m4/expat.mk
+include $(BASE_DIR)/build/m4/brotli.mk
+include $(BASE_DIR)/build/m4/freetype_bootstrap.mk
+include $(BASE_DIR)/build/m4/harfbuzz.mk
+include $(BASE_DIR)/build/m4/freetype.mk
+include $(BASE_DIR)/build/m4/fontconfig.mk
+include $(BASE_DIR)/build/m4/libass.mk
```

Comment:
This will hopefully become largely obsolete at some point due to https://github.com/libass/libass/pull/330 -- then Meson can just do a mega-build of all the projects at once via its internal subproject support.

Reply:
There are also CMake-only projects currently in there and in the future more autotools-only projects might be added.
Can a “Meson mega-build” properly parallelise CMake and “Unstable External Project” subprojects. I'd guess the former could in theory work since CMake has a Ninja-backend, but with Ninja’s lack of jobserver support I do not see how the Make-using latter could be fully parallelised together with the rest of the build.
That is assuming the warnings in Meon's docs regarding mixing build systems and the scary “experimental”/“unstable” label of the latter do not cause issues.

Reply:
In the case of cmake, Meson actually runs cmake in trace mode and extracts the information about targets, then converts them to Meson rules and builds everything directly in Meson's own build.ninja regardless.

In the Unstable External Project case, Meson has no way of knowing what make targets need to be built. So it just builds everything, and updates a `name.stamp` file on completion. Any individual target that depends on an autotools external project will have that target do a complete build of the autotools project.

Any target that does NOT depend on the external project can be parallelized with the invoked `make` command, but you are indeed correct that ninja doesn't have any jobserver support so this isn't quite so helpful.

In fact, it may even be worse. :( the invoked `make` gets `-j<nproc>` so make and ninja actually compete for jobs, and `ninja -j1` isn't respected by make at all (and it cannot even try! There's no way to probe for that info from a sub-job of ninja. Make would actually set $MAKEFLAGS based on your arguments.)

The good news is that fribidi, freetype, fontconfig and harfbuzz have meson.build support upstream, while expat and brotli (which both officially support cmake as well as autotools) both have [WrapDB](https://wrapdb.mesonbuild.com/) support via third-party maintained meson.build files hosted through Meson's official index.

So you would never have needed to touch the Unstable External Project module, only cmake, and you don't really need to do that either due to the WrapDB versions.

Reply:
> while expat and brotli (which both officially support cmake as well as autotools) both have [WrapDB](https://wrapdb.mesonbuild.com/) support via third-party maintained meson.build files hosted through Meson's official index.

The linked site indicates that the current expat wrap is version `2.29-4` *(I'm assuming this means upstream version 2.2.9, 4th iteration of the wrap for the identical upstream release)* and brotli is at `1.0.7-1`. Those upstream versions were released 2019-09-25 and 2018-10-23 respectively which is older than what JSO currently uses and with several new upstream releases since fixing security bugs.
Also JSO needs to patch some projects and their build-system to disable certain things which do not play nicely with the WASM runtime in common browser configurations. Relying on both a third-party build-system which at most updates on new releases and then our own patches at top seems more fragile than necessary.

For the above reasons, I think relying on WrapDB is not the way to go for JSO.

Reply:
> Relying on both a third-party build-system which at most updates on new releases

Build system fixes are accepted at any time, updating to unstable snapshots is not.

These are updated as time allows and when people are interested in working on them -- typical OSS. I just looked at brotli, it was totally trivial to update the version and update it to modern `[provide]` standards on the side. :)

I could take a look at expat soon, probably.

Reply:
I appreciate your work on updating them now and I'm sure the users of those wraps will do too. But in JSO, (a) we also will want to update them again sometime in the future and I'm not thrilled to maintain a bunch of full buildsystems for our dependencies and (b) JSO does not always use only releases of its subprojects. *(Currently all but libass is pointed at a release iinm, but historically and possibly in the future that won't/wasn't always the case. It definitely won't be the case for libass at least.)*